ENH: sparse: avoid densification on broadcasted elem-wise mult

perimosocordiae opened the pull request:
Fixes #5056 (again).

Now, the only way to get a dense result from `sparse.multiply(dense)` is when the sparse matrix has only one element and the dense matrix has more than one element.

rgommers commented:
Is that enough of a regression test? Maybe good to add examples in https://github.com/scipy/scipy/issues/5056#issuecomment-257148393 as a test?

rgommers commented on scipy/sparse/compressed.py:
```diff
@@ -399,42 +399,83 @@ def multiply(self, other):
+
+        other = np.atleast_2d(other)
+        # Dense matrix/array.
+        if other.ndim <= 2:
```
nitpick: can't be smaller than 2 given the ``atleast_2d`` above

rgommers commented on scipy/sparse/compressed.py:
```diff
@@ -399,42 +399,83 @@ def multiply(self, other):
-                                       ).view(np.ndarray).ravel()
-                return ret
+
+        other = np.atleast_2d(other)
```
This doesn't seem necessary and may not be quite right. If you get other objects like sparse arrays (but not captured by an if-clause above) that don't react well to ``atleast_2d``, you may break code that used to work fine with ``return np.multiply(self.todense(), other)``.

rgommers replied:
You can get this kind of thing:

    In [27]: x2 = np.atleast_2d(x)

    In [28]: x2
    Out[28]: 
    array([[ <1x1 sparse matrix of type '<class 'numpy.int64'>'
	    with 1 stored elements in Compressed Sparse Column format>]], dtype=object)

rgommers replied:
Note from the coverage report that the "anything else" branch isn't tested: https://codecov.io/gh/scipy/scipy/compare/094a004eb17c965ccf51d095a79cdc8c2c3f6c10...e59ace6ab97aecea9fdcd3c3f34aefc3d18a9080

perimosocordiae commented:
@rgommers: I reworked the logic a bit and added comments to make what's happening more clear. I also added another test matrix, so now we have full coverage on this PR. Removing the `isspmatrix` check in the unit test (see my first commit's changes) ensures that if any dense output is produced, the call to `.todense()` will throw an exception and fail the test.

We don't have a test for the "custom object" case, yet, though I'm not sure how useful it would be. This snippet:
```python
class Custom(object):
    def __rmul__(self, other):
        return 'rmul: %s' % other

x = ss.eye(2, format='csr').multiply(Custom())
```
produces `x`:
```
<2x2 sparse matrix of type '<type 'numpy.object_'>'
        with 2 stored elements in Compressed Sparse Row format>
```
with `x.data`: `array(['rmul: 1.0', 'rmul: 1.0'], dtype=object)`.

This is in line with the numpy behavior (i.e., `np.multiply(np.eye(2), Custom())`).

rgommers commented:
> Removing the `isspmatrix` check in the unit test (see my first commit's changes) ensures that if any dense output is produced, the call to `.todense()` will throw an exception and fail the test.

Indeed. Tests LGTM.

rgommers commented:
In it goes. Thanks CJ!
